cronback-lib/project_settings.rs

```diff
@@ -0,0 +1,117 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::Future;
+use moka::future::Cache;
+use moka::Expiry;
+use thiserror::Error;
+
+use crate::clients::ScopedMetadataSvcClient;
+use crate::model::ValidShardedId;
+use crate::types::{ProjectId, RequestId};
+use crate::GrpcClientFactory;
+
+type MetadataClientFactory = Arc<
+    Box<dyn GrpcClientFactory<ClientType = ScopedMetadataSvcClient> + Send>,
+>;
+
+#[derive(Error, Debug, Clone)]
+pub enum ProjectSettingError {
+    // Optimally this should wrap a `GrpcClientError` but unfortauntly, it's
+    // not `Clone` and it's not possible to make it Clone. Error being `Clone`
+    // is a requirement for the cache though.
+    #[error("failed to create metadata svc client: {0}")]
+    Client(String),
+    #[error("setting grpc called failed with code {0}")]
+    Server(#[from] tonic::Status),
+}
+
+/// A TTL-based read-through async project setting cache. A single instance of
+/// this struct caches only a single setting type.
+///
+/// Reading from the cache will return the value immedietly if it exists, and
+/// will fetch it if it doesn't exist. Notes:
+/// 1. If multiple concurrent callers request the same uncached value, only one
+///    of them will fetch it and the rest will wait for the result.
+/// 2. If the cached value exceeded the TTL, it will get refetched on the next
+///    read.
+/// 3. If the fetch from source fails, waiting callers will get fullfilled with
+///    the failed Result. However, the value will get refetched on the next read
+///    attempt (even if it didn't exceed the TTL).
+/// 4. Cache evictions happen in the background in a separate thread pool.
+pub struct ProjectSetting<V, F> {
+    client_factory: MetadataClientFactory,
+    fetcher: F,
+    cache: Cache<ValidShardedId<ProjectId>, Result<V, ProjectSettingError>>,
```

Comment:
I'm a bit torn on whether the key should be a decorated `ValidShardedId` or not. I haven't formed a strong opinion about where to enforce ValidShardedId, I'm comfortable leaving this as is, and maybe it's something we should reflect on later and see how we both feel about it.

Reply:
I like the idea of caching the non-decorated id, will switch to that.